### PR TITLE
feat: surface container artifacts via executionresult artifacts

### LINKED ISSUE #1340: python_tool() with docker tiers should surface container artifacts via ExecutionResult.artifacts
Problem

`python_tool(tier="docker")` and `python_tool(tier="docker_unsafe")` silently drop any files written inside the container. result.artifacts is always [] for docker tiers, and passing artifact_dir triggers a RuntimeWarning telling callers to switch to a local tier instead. This makes docker tiers a second-class option for agentic workflows that need the model to produce output files (plots, CSVs, generated code, etc.) — exactly the use cases where container isolation matters most.

Current behavior

python_tool() uses one-shot mode for docker tiers: a fresh SandboxSession is opened and closed inside each run_python() call.

LLMSandboxEnvironment.execute() only exports artifacts when self._session is not None (persistent context-manager mode). One-shot mode hits the guard at line 652 and skips export entirely.

artifact_dir passed with a docker tier emits a RuntimeWarning and is silently ignored (lines 1203–1211).

CapabilityPolicy.artifact_export_paths is wired and enforced for persistent sessions, but python_tool() never opens one.

Expected behavior

When artifact_export_paths is set on the policy passed to python_tool(tier="docker[_unsafe]"), files at those container paths should be copied out and returned as Artifact objects on ExecutionResult.artifacts, consistent with local tier behavior. artifact_dir with a docker tier should also be honoured (or at minimum not silently no-op).

Suggested implementation

The root cause is the mismatch between one-shot mode (used by python_tool) and the self._session is not None guard on the artifact export path (line 741). The fix: make python_tool() open a persistent LLMSandboxEnvironment context when a docker tier is selected and artifact export is configured, instead of delegating to one-shot mode.

1. In python_tool(), detect docker tier + artifact_export_paths set (or artifact_dir provided).
2. Open LLMSandboxEnvironment as a context manager once at tool construction time, keeping _session alive across run_python() calls.
3. The returned MelleaTool owns the container lifecycle — use contextlib.ExitStack or a __del__ finalizer to ensure teardown.
4. Wire artifact_dir as the host-side destination for copied-out files, replacing the tempfile.gettempdir() fallback in the export loop (lines 744–761).
5. Remove the RuntimeWarning for artifact_dir + docker tier once it is supported.

The persistent session also improves performance — avoids repeated container cold-start and package reinstall across multiple tool calls in a ReACT loop.

Files of interest

mellea/stdlib/tools/interpreter.py — python_tool() (line 1039), LLMSandboxEnvironment.execute() (line 652), artifact export loop (line 737)

mellea/stdlib/tools/execution_policy.py — CapabilityPolicy.artifact_export_paths field

### PR DESCRIPTION
**Pull request by AngeloDanducci**
# Pull Request

## Issue
Fixes #1340 

## Description

surface container artifacts via ExecutionResult.artifacts - users should not need to dig into their local container to pull artifacts created during the sandbox use via python tool

### Testing
- [x] Tests added to the respective file if code was changed
- [ ] New code has 100% coverage if code was added
- [ ] Ensure existing tests and github automation passes (a maintainer will kick off the github automation when the rest of the PR is populated)

### Attribution
- [x] AI coding assistants used

---

### Adding a new component, requirement, sampling strategy, or tool?

If your PR adds or modifies one of the types below, check the matching box. A checklist of type-specific review items will be posted as a comment.

- [ ] Component
- [ ] Requirement
- [ ] Sampling Strategy
- [ ] Tool

NOTE: Please ensure you have an issue that has been acknowledged by a core contributor and routed you to open a pull request against this repository. Otherwise, please open an issue before continuing with this pull request.

### CODE REVIEW

File: mellea/stdlib/tools/interpreter.py [hunk 1]
```diff
@@ -745,26 +753,47 @@ def execute(self, code: str, timeout: int | None = None) -> ExecutionResult:
 
             artifacts: list[Artifact] = []
             export_dirs: list[Path] = []
+            # When a host export_dir is configured, artifacts persist there for
+            # the caller and are NOT auto-cleaned; otherwise they land in a
+            # randomized system-tempdir subdir removed once the result is GC'd.
+            caller_owns_exports = self._export_dir is not None
```
Comment (markstur) [inline mellea/stdlib/tools/interpreter.py:769]: I'm not 100% sure my thoughts here are correct all the way through, but it seems to me that there is a corner case where _export_dir is falsey Path("") and would behave badly.  Also I notice this because in reading the code my opinion was that I'd remove caller_owns_exports and just use "if self._export_dir" in places where we decide what to do or not do with _export_dir.   

So the opinion on readability of caller_owns_exports is totally up to you, but  I think you should test with an empty or falsey _export_dir.  It seems like it results in a weird use of system temp but caller-owns true.  I haven't tested it though, so I wouldn't be surprised if you find the edge case is somehow a non-issue.

**Review by markstur — COMMENTED** [review]
Looks pretty good.  I'm just questioning what a falsey export_dir would do (e.g. Path("")).  This is probably just me trying to use falsey instead of "is not None" but if this is real possibility then I think it results in a weird use of system temp while setting caller-owns to True.  Needs testing.

The readability comment... is probably more author preference so feel free to keep the current style if that's your preference.  Just felt to me like the conditions were all essentially "if _export_dir" do export dir things.

**Review by ajbozarth — COMMENTED** [review]
Some feedback from Claude. Approach is sound and tests are thorough. Two non-blocking items inline; not gating on either.

File: mellea/stdlib/tools/interpreter.py [hunk 2]
```diff
@@ -1236,6 +1329,10 @@ def run_python(code: str) -> ExecutionResult:
             else ""
         ) + code
 
+        if persistent_env is not None:
+            # Docker tier with artifact export: reuse the open container session.
+            return persistent_env.execute(patched_code)
```
Comment (ajbozarth) [inline mellea/stdlib/tools/interpreter.py:1353]: Add a stale-artifact caveat to the `python_tool` Note block. The container persists and `artifact_export_paths` are fixed, so a later successful run that doesn't touch an export path will still `copy_out` a file left by an earlier run and surface it as the current run's output. Consistent with persistent local `artifact_dir`, but the reused container makes it easy to hit in a ReACT loop — callers should know.

File: mellea/stdlib/tools/interpreter.py [hunk 3]
```diff
@@ -1132,8 +1180,19 @@ def python_tool(
 
     Raises:
         ImportError: If `MelleaTool` cannot be imported (should not happen in
-            a normal mellea installation).
+            a normal mellea installation), or, for a docker tier with
+            `artifact_export_paths` set, if `llm-sandbox` is not installed —
+            the persistent container session is opened at construction time.
         ValueError: If any entry in `packages` is empty or begins with `-`.
+        RuntimeError: If a docker tier with `artifact_export_paths` set cannot
```
Comment (ajbozarth) [inline mellea/stdlib/tools/interpreter.py]: `SandboxSession.open()` against an unreachable daemon raises a `docker.errors.*`/llm-sandbox exception, not `RuntimeError`, so this `Raises:` type is misleading. Suggest generalizing:

```suggestion
        RuntimeError: If a docker tier with `artifact_export_paths` set cannot
            open its container session at construction time — the underlying
            docker/llm-sandbox error (e.g. the Docker daemon is unreachable)
            propagates unchanged.
```

Comment (AngeloDanducci) [reply mellea/stdlib/tools/interpreter.py]: added

File: mellea/stdlib/tools/interpreter.py (hunk 2 above)
Comment (AngeloDanducci) [reply mellea/stdlib/tools/interpreter.py:1353]: added

File: mellea/stdlib/tools/interpreter.py (hunk 1 above)
Comment (AngeloDanducci) [reply mellea/stdlib/tools/interpreter.py:769]: Took another look at this and did a slight change.

Any empty/cwd path ( Path("") or Path("."), both stringify to "." ) should be normalized to None at construction. So downstream, `caller_owns_exports = self._export_dir is not None` is now `False` for those inputs, and artifacts correctly fall through to the auto-cleaned system tempdir.

**Review by markstur — APPROVED** [review]
Thanks.  LGTM

**Review by ajbozarth — APPROVED** [review]
